drop-transfer/src/check.rs

```diff
@@ -7,51 +7,91 @@ use tokio_util::sync::CancellationToken;
 use crate::{protocol, service::State, tasks::AliveGuard, IncomingTransfer, Transfer};
 
 pub(crate) fn spawn(
+    refresh_trigger: tokio::sync::watch::Receiver<()>,
     state: Arc<State>,
     xfer: Arc<IncomingTransfer>,
     logger: Logger,
     guard: AliveGuard,
     stop: CancellationToken,
-) -> tokio::task::JoinHandle<()> {
+) {
     let id = xfer.id();
-    let job = run(state, xfer, logger.clone());
 
     tokio::spawn(async move {
         let _guard = guard;
+        let mut refresh_trigger = refresh_trigger;
+
+        let logger_clone = logger.clone();
+        let task = async move {
+            let logger = logger_clone;
+            loop {
+                let cf = run(&state, xfer.clone(), &logger).await;
+
+                if cf.is_break() {
+                    break;
+                }
+
+                #[allow(clippy::let_underscore_must_use)]
+                let _ = refresh_trigger.changed().await;
+            }
+        };
 
         tokio::select! {
             biased;
 
             _ = stop.cancelled() => {
-                debug!(logger, "Check job stop for transfer: {id}");
+                debug!(&logger, "stop checking job for: {}", id);
             },
-            _ = job => (),
+            _ = task => ()
         }
-    })
+    });
 }
 
-async fn run(state: Arc<State>, xfer: Arc<IncomingTransfer>, logger: Logger) {
-    loop {
-        tokio::time::sleep(drop_config::ALIVE_CHECK_INTERVAL).await;
+async fn run(state: &State, xfer: Arc<IncomingTransfer>, logger: &Logger) -> ControlFlow<()> {
+    let mut remove_xfer = false;
+    for delay in std::iter::once(Duration::from_secs(0)).chain(drop_config::RETRY_INTERVALS) {
+        debug!(
+            logger,
+            "Incoming transfer job started for {}, will sleep for {:?}",
+            xfer.id(),
+            delay
+        );
+        tokio::time::sleep(delay).await;
 
         if !state.transfer_manager.is_incoming_alive(xfer.id()).await {
+            remove_xfer = true;
             break;
         }
 
-        if make_request(&state, &xfer, &logger).await.is_break() {
+        if make_request(state, &xfer, logger).await.is_break() {
+            remove_xfer = true;
             break;
         }
+
+        debug!(
+            logger,
+            "Client job will restart in {:?} seconds for {}",
+            delay,
+            xfer.id()
+        );
```

Comment:
Give that there's a debug log above, maybe this one is unneeded, or at the very least shouldn't proc on the last iteration since it won't actually restart, which could possibly cause confusion when someone reads the logs

Reply:
actually it was not necessary anymore, correct